### PR TITLE
[FIRRTL] Move transformation into a transformation library

### PR DESCRIPTION
**Pull request by youngar**
We will eventually adding some more transformation passes to FIRRTL, so
I thought that it would be a good idea to move them into their own
library. A couple upgrades were made on the way.

ODS pass registration was moved into a headfile, to be more inline with
MLIR.  This means we no long have to manually create a pass registration
function.  There seems to be two common choices for naming the pass
registration, the second one was taken:

```c++
void registerFIRRTLPasses();
void firrtl::registerPasses();
```

Without having to worry about pass registration, I was able to delete
DialectRegistration.cpp.

One thing which was not totally clear was where the `Passes.h` header
file should live: in `FIRRTL` or `FIRRTL/Transfoms`.  There seems to be
a pretty even split in MLIR.  I put Passes.h `FIRRTL/Transforms` since
users are required to link against the Transforms library to use the
functions.

The pass `lower-firrtl-types` was renamed to `firrtl-lower-types` under
the idea that all dialect specific passes should be namespaced by the
dialect. We could also have also named the pass
`firrtl-lower-firrtl-types` if that makes more sense.

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/Transforms/Passes.td [hunk 1]
```diff
@@ -4,17 +4,17 @@
...
 
-#ifndef CIRCT_DIALECT_FIRRTL_PASSES_TD
-#define CIRCT_DIALECT_FIRRTL_PASSES_TD
+#ifndef CIRCT_DIALECT_FIRRTL_TRANSFORMS_PASSES_TD
+#define CIRCT_DIALECT_FIRRTL_TRANSFORMS_PASSES_TD
 
 include "mlir/Pass/PassBase.td"
 
-def LowerFIRRTLTypes : Pass<"lower-firrtl-types", "firrtl::FModuleOp"> {
+def LowerFIRRTLTypes : Pass<"firrtl-lower-types", "firrtl::FModuleOp"> {
```
Comment (mikeurbach) [inline include/circt/Dialect/FIRRTL/Transforms/Passes.td:12]: The naming change sounds good to me. The previous name was in line with the old `lower-to-rtl` pass, but the new name seems better in the long run..

**Review by mikeurbach — COMMENTED** [review]
Thanks for doing this, LGTM if the build passes.

**Comment by lattner** [comment]
I would only create `include/circt/Dialect/FIRRTL/Transforms/` if we plan to have other header files in that directory.  If you have a clear sightline on that, adding the directory makes sense.  Otherwise, I'd just name the file "circt/Dialect/FIRRTL/Transforms.h"

**Comment by lattner** [comment]
Patch LGTM with that thought!

**Comment by youngar** [comment]
Thanks for that feedback!  I think that passes tend to only have a single header file, so I'll remove the nesting.

**Comment by youngar** [comment]
Moved `Passes.h` and `Passes.td` out of the transforms directory.  It seems a little weird to me to not have a 1-to-1 match between the `include` tree and the `lib` tree, but it seems fairly normal in LLVM.

**Review by lattner — APPROVED** [review]

**Pull request merged**